### ISSUE TITLE
the option --force-recreate with the command docker-compose didn't work

### ISSUE THREAD

**Issue by cloedy** [docker-library/docs#1287]
```
FROM php:7.2-fpm
RUN apt-get update && apt-get install -y \
        libfreetype6-dev \
        libjpeg62-turbo-dev \
        libpng-dev \
	libmemcached-dev \
    && docker-php-ext-install -j$(nproc) iconv \
    && docker-php-ext-configure gd --with-freetype-dir=/usr/include/ --with-jpeg-dir=/usr/include/ \
    && docker-php-ext-install -j$(nproc) gd  \
    && docker-php-ext-install -j$(nproc) pdo \
    && docker-php-ext-install -j$(nproc) pdo_mysql \
    && docker-php-ext-install -j$(nproc) pcntl \
    && docker-php-ext-install -j$(nproc) bcmath \
    && docker-php-ext-install -j$(nproc) mysqli \
	&& pecl install redis-4.0.1 \
    && docker-php-ext-enable redis \
	&& pecl install memcached \
    && docker-php-ext-enable memcached \
	&& echo "alias ll='ls -l'" >> $HOME/.bashrc
```
here is my php dockerfile (memcached install is added after this execute)
```
docker-compose up -d --build
```
then I update the Dockerfile to install memcached extension, but after the command
```
docker-compose up -d --build --force-recreate 
```
then I met a problem
```
pecl/redis is already installed and is the same as the released version 4.0.1
```
that is the thing I installed last time , so I doubt the option --force-recreate didn't really working😩
it use the builded thing
how can I figure out the problem, without rm the image and container, any suggestion will be appreciated

**Comment by wglambert** [comment]
Can't reproduce an issue with `--force-recreate` , you would need to troubleshoot the error with your Dockerfile, you could ask [the Docker Community Forums](https://forums.docker.com/), [the Docker Community Slack](https://blog.docker.com/2016/11/introducing-docker-community-directory-docker-community-slack/), or [Stack Overflow](https://stackoverflow.com/search?tab=newest&q=docker).


docker-compose.yml
```yaml
version: '3'

services:
   php:
     image: php:test
```
```console
$ docker build . -t php:test
Sending build context to Docker daemon  3.584kB
Step 1/2 : FROM php:7.2-fpm
 ---> da6c31a7f6b1
Step 2/2 : RUN apt-get update && apt-get install -y         libfreetype6-dev         libjpeg62-turbo-dev         libpng-dev     libmemcached-dev     && docker-php-ext-install -j$(nproc) iconv     && docker-php-ext-configure gd --with-freetype-dir=/usr/include/ --with-jpeg-dir=/usr/include/     && docker-php-ext-install -j$(nproc) gd      && docker-php-ext-install -j$(nproc) pdo     && docker-php-ext-install -j$(nproc) pdo_mysql     && docker-php-ext-install -j$(nproc) pcntl     && docker-php-ext-install -j$(nproc) bcmath     && docker-php-ext-install -j$(nproc) mysqli  && pecl install redis-4.0.1     && docker-php-ext-enable redis && pecl install memcached     && docker-php-ext-enable memcached         && echo "alias ll='ls -l'" >> $HOME/.bashrc
 ---> Using cache
 ---> e32d0294cb0f
Successfully built e32d0294cb0f
Successfully tagged php:test
```
```console
$ docker-compose up -d
Creating network "docs-1287_default" with the default driver
Creating docs-1287_php_1 ... done
$ docker ps
CONTAINER ID        IMAGE               COMMAND                  CREATED             STATUS              PORTS               NAMES
138e9e23757d        php:test            "docker-php-entrypoi…"   16 seconds ago      Up 14 seconds       9000/tcp            docs-1287_php_1
$ docker-compose up -d --force-recreate
Recreating docs-1287_php_1 ... done
$ docker ps
CONTAINER ID        IMAGE               COMMAND                  CREATED             STATUS              PORTS               NAMES
adbd532cad4f        php:test            "docker-php-entrypoi…"   3 seconds ago       Up 1 second         9000/tcp            docs-1287_php_1
$ docker-compose logs
Attaching to docs-1287_php_1
php_1  | [27-Aug-2018 17:20:21] NOTICE: fpm is running, pid 1
php_1  | [27-Aug-2018 17:20:21] NOTICE: ready to handle connections
```

**Comment by yosifkit** [comment]
Random thought, did you put both `image` and `build` in your compose?

```yaml
services:
  php:
    image: php:7.2-fpm
    build: ./
```

This would cause your image to be tagged as `php:7.2-fpm` and then the next time you built it would be `FROM` your previous image.

**Comment by wglambert** [comment]
Doesn't appear to be any unexpected behavior

docker-compose.yml:
```yaml
version: '3'

services:
   php:
     image: php:test
     build: ./
```

```console
$ docker ps
CONTAINER ID        IMAGE               COMMAND             CREATED             STATUS              PORTS               NAMES
$ docker-compose up -d
Creating docs-1287_php_1 ... done
$ docker-compose logs
Attaching to docs-1287_php_1
php_1  | [27-Aug-2018 20:23:43] NOTICE: fpm is running, pid 1
php_1  | [27-Aug-2018 20:23:43] NOTICE: ready to handle connections
$ docker ps
CONTAINER ID        IMAGE               COMMAND                  CREATED             STATUS              PORTS               NAMES
24fb92f3b856        php:test            "docker-php-entrypoi…"   25 seconds ago      Up 23 seconds       9000/tcp            docs-1287_php_1
$ docker-compose up -d --force-recreate
Recreating docs-1287_php_1 ... done
$ docker-compose logs
Attaching to docs-1287_php_1
php_1  | [27-Aug-2018 20:24:22] NOTICE: fpm is running, pid 1
php_1  | [27-Aug-2018 20:24:22] NOTICE: ready to handle connections
$ docker ps
CONTAINER ID        IMAGE               COMMAND                  CREATED             STATUS              PORTS               NAMES
8ed9575d1b27        php:test            "docker-php-entrypoi…"   16 seconds ago      Up 15 seconds       9000/tcp            docs-1287_php_1
```

**Comment by cloedy** [comment]
ah. it's my fault to make such a horrible description
here's my docker-compose.yml
```
version: "3.7"

networks:
    local-net:
        driver: bridge

services:
    nginx:
        build:
            context: ./nginx
        container_name: nginx.dev
        ports:
            - 10000:80
        volumes:
            - ./nginx/blog.conf:/etc/nginx/conf.d/blog.conf
            - ./blog:/www/blog
        networks:
            - local-net
        tty: true
        links:
            - "php:phper"
    php:
        build:
            context: ./php
        container_name: php.dev
        ports:
            - 9002:9000
        networks:
            - local-net
        volumes:
            - ./blog:/www/blog
    mysql:
        build:
            context: ./mysql
        container_name: mysql.dev
        volumes:
            - ./mysql/init.d:/docker-entrypoint-initdb.d
        ports:
            - 33060:3306
        networks:
            - local-net
        environment:
            MYSQL_ROOT_PASSWORD: admin
        command: --default-authentication-plugin=mysql_native_password
```
and the php Dockerfile is like this
```
  1 FROM php:7.2-fpm
  2 RUN apt-get update && apt-get install -y vim \
  3         libfreetype6-dev \
  4         libjpeg62-turbo-dev \
  5         libpng-dev \
  6         libmemcached-dev \
  7     && docker-php-ext-install -j$(nproc) iconv \
  8     && docker-php-ext-configure gd --with-freetype-dir=/usr/include/ --with-jpeg-dir=/usr/include/ \
  9     && docker-php-ext-install -j$(nproc) gd  \
 10     && docker-php-ext-install -j$(nproc) pdo \
 11     && docker-php-ext-install -j$(nproc) pdo_mysql \
 12     && docker-php-ext-install -j$(nproc) pcntl \
 13     && docker-php-ext-install -j$(nproc) bcmath \
 14     && docker-php-ext-install -j$(nproc) mysqli \
 15     && pecl install redis-4.1.1 \
 16     && docker-php-ext-enable redis \
 17     && echo "alias ll='ls -l'" >> $HOME/.bashrc
 18
 19 WORKDIR /www/blog
```
my directory is like this 
```
.
├── blog
├── docker-compose.yml
├── mysql
│   ├── Dockerfile
│   └── init.d
│       └── init.sql
├── nginx
│   ├── Dockerfile
│   ├── blog
│   └── blog.conf
└── php
    └── Dockerfile
```
first time I ran 
```
docker-compose up -d --build
```
and it built success with no doubt, then i realized that i missing the memcached extension (maybe something other extension in future) for php, so the php Dockerfile became like this
```
  1 FROM php:7.2-fpm
  2 RUN apt-get update && apt-get install -y vim \
  3         libfreetype6-dev \
  4         libjpeg62-turbo-dev \
  5         libpng-dev \
  6         libmemcached-dev \
  7     && docker-php-ext-install -j$(nproc) iconv \
  8     && docker-php-ext-configure gd --with-freetype-dir=/usr/include/ --with-jpeg-dir=/usr/include/ \
  9     && docker-php-ext-install -j$(nproc) gd  \
 10     && docker-php-ext-install -j$(nproc) pdo \
 11     && docker-php-ext-install -j$(nproc) pdo_mysql \
 12     && docker-php-ext-install -j$(nproc) pcntl \
 13     && docker-php-ext-install -j$(nproc) bcmath \
 14     && docker-php-ext-install -j$(nproc) mysqli \
 15     && pecl install redis-4.1.1 \
 16     && docker-php-ext-enable redis \
 17     && pecl install memcached \
 18     && docker-php-ext-enable memcached \
 19     && echo "alias ll='ls -l'" >> $HOME/.bashrc
 20
 21 WORKDIR /www/blog
```
just added line 17 && 18, then I ran the command
```
docker-compose up -d --build --force-recreate php
```
after a few seconds then I got some tips like this 
```
pecl/redis is already installed and is the same as the released version 4.1.1
install failed
ERROR: Service 'php' failed to build: The command '/bin/sh -c apt-get update && apt-get install -y vim         libfreetype6-dev         libjpeg62-turbo-dev         libpng-dev      libmemcached-dev     && docker-php-ext-install -j$(nproc) iconv     && docker-php-ext-configure gd --with-freetype-dir=/usr/include/ --with-jpeg-dir=/usr/include/     && docker-php-ext-install -j$(nproc) gd      && docker-php-ext-install -j$(nproc) pdo     && docker-php-ext-install -j$(nproc) pdo_mysql     && docker-php-ext-install -j$(nproc) pcntl     && docker-php-ext-install -j$(nproc) bcmath     && docker-php-ext-install -j$(nproc) mysqli  && pecl install redis-4.1.1     && docker-php-ext-enable redis  && echo "alias ll='ls -l'" >> $HOME/.bashrc' returned a non-zero code: 1
```
what I want to do is to add a php-memcached extension (configuration changed), then I added the option param "--force-recreate" to rebuild it, but it didn't work as I wished as we see. that confused me if the recreate processing use the container built last time? (while "--force-recreate" means  Recreate containers even if their configuration and image haven't changed.)
